### PR TITLE
timealarms: single alarm per address

### PR DESCRIPTION
**Pull request by go2bodhi**

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -1250,6 +1250,20 @@ dependencies = [
...
+ "cosmwasm-std",
+ "platform",
+ "schema",
+ "sdk",
+ "serde",
+ "serde-json-wasm 0.5.0",
+ "thiserror",
+ "timealarms",
+]
+
```
Comment (gmanev7) [inline Cargo.lock:1266]: where is the code of this new package?

File: contracts/timealarms/src/alarms.rs [hunk 2]
```diff
@@ -30,28 +41,42 @@ impl TimeAlarms {
...
-        let dispatcher = Self::TIME_ALARMS.notify(storage, dispatcher, ctime, max_count)?;
+        let dispatcher = self
+            .time_alarms
+            .alarms_selection(storage, ctime)
+            .take(max_count.try_into()?)
+            .enumerate()
+            .try_fold(
+                OracleAlarmDispatcher::new(),
+                |dispatcher, (id, alarm)| -> Result<OracleAlarmDispatcher, ContractError> {
+                    dispatcher.send_to(id.try_into()?, alarm?.0)
```
Comment (gmanev7) [inline contracts/timealarms/src/alarms.rs:62]: I believe we should stop using generated Id-s. Any constant would suffice.

File: contracts/timealarms/src/dispatcher.rs [hunk 3]
```diff
@@ -25,16 +26,17 @@ impl OracleAlarmDispatcher {
             emitter: Emitter::of_type(EVENT_TYPE),
         }
     }
-}
-impl AlarmDispatcher for OracleAlarmDispatcher {
-    fn send_to(mut self, id: Id, addr: Addr) -> Result<Self, AlarmError> {
+
+    pub fn send_to(mut self, id: Id, addr: Addr) -> Result<Self, ContractError> {
```
Comment (gmanev7) [inline contracts/timealarms/src/dispatcher.rs:30]: why do we need it `pub`?

Comment (go2bodhi) [reply contracts/timealarms/src/dispatcher.rs:30]: earlier it was in the impl of a trait, so it was `pub`. It's used outside in the `timealarms::alarms`, don't see the problem.

File: contracts/timealarms/src/alarms.rs (hunk 2 above)
Comment (gmanev7) [reply contracts/timealarms/src/alarms.rs:62]: @go2bodhi I do not see it resolved

Comment (go2bodhi) [reply contracts/timealarms/src/alarms.rs]: forgot to commit

File: packages/time-oracle/src/migrate_v1.rs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        alarms.add(storage, addr3, t3).unwrap();
+        alarms.add(storage, addr4, t4).unwrap();
+
+        // multiple alarms per address(5) + index(5) + next_id(1)
+        assert_eq!(11, storage.range(None, None, Order::Ascending).count());
+
+        alarms.migrate(storage).unwrap();
+
+        // single alarm per address(4) + index(4)
+        assert_eq!(8, storage.range(None, None, Order::Ascending).count());
```
Comment (gmanev7) [inline packages/time-oracle/src/migrate_v1.rs:156]: please add more checks on the alarms post-migration, e.g. what alarms for what addresses are there

File: packages/time-oracle/src/migrate_v1.rs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+use sdk::{
+    cosmwasm_std::{Addr, Order, StdError, Storage, Timestamp},
+    cw_storage_plus::{Index, IndexList, IndexedMap, Item, MultiIndex},
+};
+
+use super::Alarms;
+use crate::AlarmError;
+
+type TimeSeconds = u64;
```
Comment (gmanev7) [inline packages/time-oracle/src/migrate_v1.rs:11]: if it has not been changed then use the one from v1

File: contracts/timealarms/src/alarms.rs [hunk 6]
```diff
@@ -30,28 +42,52 @@ impl TimeAlarms {
...
+    pub fn migrate_v1(storage: &mut dyn Storage) -> Result<(), ContractError> {
+        AlarmsOld::new(
+            Self::ALARMS_NAMESPACE,
+            Self::ALARMS_IDX_NAMESPACE,
+            "alarms_next_id",
+        )
+        .migrate(storage)?;
+
+        Ok(())
     }
```
Comment (gmanev7) [inline contracts/timealarms/src/alarms.rs:91]: let's not mix the migration code with the rest

The aim is to locate it in module(s) that are going to be removed once the migration is over.

File: contracts/timealarms/Cargo.toml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "timealarms"
-version = "0.2.4"
+version = "0.3.0"
```
Comment (gmanev7) [inline contracts/timealarms/Cargo.toml:3]: there is no API change, so let's stay with the same major version

Comment (gmanev7) [reply contracts/timealarms/Cargo.toml:3]: since it introduces a response so disregard

**Pull request merged**